### PR TITLE
fix kubectl --local flag bug across all commands

### LINKED ISSUE #93188: kubectl --local is not quite local across all the commands
This situation was discovered when I wanted to remove deprecated functionality in https://github.com/kubernetes/kubernetes/pull/90243. Unfortunately, due to how we are constructing client in almost every kubectl command this removal ended up with the following bug https://github.com/kubernetes/kubernetes/issues/90074.

The reason for that is that in each command's `Complete` method we create a client in one of the following ways:
```go
o.Client, err = batchv1client.NewForConfig(clientConfig)
if err != nil {
	return err
}
```
```go
dynamicClient, err := f.DynamicClient()
if err != nil {
	return err
}
```
```go
discoveryClient, err := f.ToDiscoveryClient()
if err != nil {
	return err
}
```
The problem is that with the deprecated code from #90243 removed all of the above calls will fail when creating a client with:
```
error: Missing or incomplete configuration info.  Please point to an existing, complete config file:

  1. Via the command-line flag --kubeconfig
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config
```
which does not happen if we silently default the server to `http://localhost:8080` in deprecated `getDefaultServer`. 

There are two possible approaches to this problem:
1. Go through each and every single command and add conditions for `--local` flag and not to create clients when it is specified.
2. Provide a smart mechanism allowing to inject dependencies (in this particular case, we are talking about clients).

During SIG-CLI call on July 15th, we agreed that solution 1 is a short-term and rather hacky approach to the problem and we would like to see this being approached as described in no. 2. 

Any proposals to this should be discussed with @soltysh or during one of sig-cli calls.

### PR DESCRIPTION
**Pull request by SaiHarshaK**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Kubectl "local" flag won't ask for kubeconfig after deprecated functionality in [#90243](https://github.com/kubernetes/kubernetes/pull/90243) is removed.

**Which issue(s) this PR fixes**:
Fixes #93188

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/clientcmd/client_config.go [hunk 1]
```diff
@@ -37,23 +37,14 @@ import (
...
 	// DefaultClientConfig represents the legacy behavior of this package for defaulting
 	// DEPRECATED will be replace
 	DefaultClientConfig = DirectClientConfig{*clientcmdapi.NewConfig(), "", &ConfigOverrides{
 		ClusterDefaults: ClusterDefaults,
 	}, nil, NewDefaultClientConfigLoadingRules(), promptedCredentials{}}
 )
 
-// getDefaultServer returns a default setting for DefaultClientConfig
-// DEPRECATED
-func getDefaultServer() string {
```
[the remark this answers is not in the document]
Comment (SaiHarshaK) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:50]: I'll revert the changes to this file

[the remark this answers is not in the document]
Comment (SaiHarshaK) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go]: Ill squash the commits?

File: pkg/kubectl/cmd/convert/convert.go [hunk 2]
```diff
@@ -118,9 +118,11 @@ func (o *ConvertOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) (err er
 	}
 	o.builder = f.NewBuilder
 
-	o.Namespace, _, err = f.ToRawKubeConfigLoader().Namespace()
-	if err != nil {
-		return err
+	if !o.local {
```
Comment (soltysh) [inline pkg/kubectl/cmd/convert/convert.go:121]: Although this solves the problem, that's not the right approach. See what I wrote in https://github.com/kubernetes/kubernetes/issues/93188, towards the bottom. You picked the hacky solution which we don't want to happen. The entire sig including the tech leads would like to see a more general solution, rather this one off.

**Review by soltysh — CHANGES_REQUESTED** [review]
/hold

**Comment by soltysh** [comment]
/close
Let's discuss possible approaches in #93188, and only open PRs once we have a general consensus about it.

**Pull request closed without merging**